evm/vm/forks/sharding/windback_worker.py

```diff
@@ -0,0 +1,212 @@
+            except StopIteration:
+                return None
+            head_collation_hash = head_header.hash
+            if not self.chain_validity[head_collation_hash]:
+                continue
+            current_collation_hash = head_collation_hash
+            for _ in range(self.vmc.config['WINDBACK_LENGTH'] + 1):
+                # if time is up
+                if self.is_time_up:
+                    return head_collation_hash
+                if current_collation_hash == GENESIS_COLLATION_HASH:
+                    break
+
+                # if a collation is an invalid head, set its descendants in the
+                # chain as invalid heads, and skip this chain
+                is_chain_of_current_collation_invalid = (
+                    (not self.chain_validity[current_collation_hash]) or
+                    (not self.collation_validity_cache.get(current_collation_hash, True))
+                )
+                if is_chain_of_current_collation_invalid:
+                    self.chain_validity[current_collation_hash] = False
+                    for collation_hash in self.chain_collations[head_collation_hash]:
+                        self.chain_validity[collation_hash] = False
+                    break
+                # process current collation  ##################################
+                # if the collation is checked before, then skip it
+                if current_collation_hash not in self.collation_validity_cache:
+                    descendants = copy.deepcopy(
+                        self.chain_collations[head_collation_hash]
+                    )
+                    coro = self.process_collation(
+                        head_collation_hash,
+                        current_collation_hash,
+                        descendants,
+                    )
+                    task = asyncio.ensure_future(coro)
+                    asyncio.wait(task, timeout=self.YIELDED_TIME)
+                    self.set_collation_task(head_collation_hash, current_collation_hash, task)
+
+                # clean up the finished tasks, yield CPU to tasks
+                if len(self.unfinished_verifying_tasks) != 0:
+                    await asyncio.wait(self.unfinished_verifying_tasks, timeout=self.YIELDED_TIME)
+                    self.unfinished_verifying_tasks = clean_done_task(
+                        self.unfinished_verifying_tasks
+                    )
+                # yield the CPU to other coroutine
+                await asyncio.sleep(self.YIELDED_TIME)
+
+                current_collation_hash = self.vmc.get_collation_parent_hash(
+                    self.get_shard_id(),
+                    current_collation_hash,
+                )
+
+            # when `WINDBACK_LENGTH` or GENESIS_COLLATION is reached
+            while self.chain_validity[head_collation_hash]:
+                # if time is up
+                if self.is_time_up:
+                    return head_collation_hash
```

Comment:
Tasks which are still running here should probably be canceled somehow.

Reply:
look into the `CancelToken` in the `p2p` module (we can extract it to it's own library if it's needed in another codebase.

Reply:
@jannikluhn Originally I thought, since this instance will be reused when the next `guess_head` is called, we can finished that task at that time?
@pipermerriam Thank you for the direction. I will look into it.